### PR TITLE
WebScout fixes

### PR DESCRIPTION
**Pull request by ScottLouvau**
- Result.GetRule now looks up by RuleId (after RuleIndex or Guid)
- ToolComponent builds lookup cache by RuleId and Guid to prevent performance problems looking up every Result's Rule.
- Baselining now persists Run.Tool.Driver.Rules referenced by Results via RuleId properly.
- SarifToCsv: Match Sarif SDK Newtonsoft version.
- SarifToCsv: Make use of DeferredOM configurable, as workaround if needed.
- Fix Deferred OM to detect and handle files with BOMs.
- Fix broken unit tests (unrelated to these changes)

### CODE REVIEW

File: src/Sarif/Core/Result.cs [hunk 1]
```diff
@@ -73,18 +73,21 @@ public ReportingDescriptor GetRule(Run run = null)
                     return GetRuleByIndex(rules, this.Rule.Index);
                 }
 
-                // Look up by this.RuleDescriptor, Guid, if present
-                if (!string.IsNullOrEmpty(this.Rule?.Guid))
+                // Look up by this.RuleDescriptor.Guid, if present
```
Comment (lgolding) [inline src/Sarif/Core/Result.cs:76]: >RuleDescriptor [](start = 35, length = 14)

".RuleDescriptor" => ".Rule" (but per Michael, and I agree, we don't need to look up by GUID)

File: src/Sarif/Core/Result.cs [hunk 2]
```diff
@@ -73,18 +73,21 @@ public ReportingDescriptor GetRule(Run run = null)
...
-                        foreach (ReportingDescriptor rule in rules)
-                        {
-                            if (rule.Guid == this.Rule.Guid) { return rule; }
-                        }
-
-                        throw new ArgumentException($"ReportingDescriptorReference referred to Guid {this.Rule.Guid}, which was not found in toolComponent.Rules.");
-                    }
+                    ReportingDescriptor rule = component.GetRuleByGuid(this.Rule.Guid);
+                    if (rule != null) { return rule; }
+                    throw new ArgumentException($"ReportingDescriptorReference referred to Guid {this.Rule.Guid}, which was not found in toolComponent.Rules.");
```
Comment (lgolding) [inline src/Sarif/Core/Result.cs:81]: >throw new ArgumentException($"ReportingDescriptorReference referred to Guid {this.Rule.Guid}, which was not found in toolComponent.Rules.") [](start = 20, length = 139)

I don't think we should throw because the SARIF is invalid. Just don't return a `ReportingDescriptor` if you can't find one (or, in this case, fall through to the next block). Comment applies to the ruleId lookup as well.

File: src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs [hunk 3]
```diff
@@ -268,23 +268,23 @@ private List<ExtractedResult> ExtractResultsFromRuns(IEnumerable<Run> sarifRuns)
...
-                        }
-                        else
-                        {
-                            result.RuleIndex = reportingDescriptors.Count;
-                            reportingDescriptors[rule] = reportingDescriptors.Count;
+                        result.RuleIndex = ruleIndex;
+                        result.RuleId = ruleId;
+                    }
+                    else
+                    {
```
Comment (lgolding) [inline src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs:280]: >{ [](start = 20, length = 1)

It looks like you don't set `result.RuleId` in the `else` branch. If so, a test hole as well.

File: src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs [hunk 4]
```diff
@@ -268,23 +268,23 @@ private List<ExtractedResult> ExtractResultsFromRuns(IEnumerable<Run> sarifRuns)
...
+                        ReportingDescriptor rule = result.GetRule(resultPair.Run);
 
-                            run.Tool.Driver.Rules = run.Tool.Driver.Rules ?? new List<ReportingDescriptor>();
-                            run.Tool.Driver.Rules.Add(rule);
-                        }
+                        result.RuleIndex = reportingDescriptors.Count;
+                        reportingDescriptors[ruleId] = reportingDescriptors.Count;
+
+                        run.Tool.Driver.Rules = run.Tool.Driver.Rules ?? new List<ReportingDescriptor>();
```
Comment (lgolding) [inline src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs:286]: >= [](start = 46, length = 1)

Are we building SDK with C# 8 now? If so can change this to, `run.Tool.Driver.Rules ??= new List<ReportingDescriptor>();`

File: src/Sarif/Core/ToolComponent.cs [hunk 5]
```diff
@@ -1,13 +1,60 @@
...
+                if (r.Guid != null) { _cachedRulesByGuid[r.Guid] = r; }
+            }
+        }
+
+        public ReportingDescriptor GetRuleById(string ruleId)
+        {
+            ReportingDescriptor rule = null;
+
+            // Build lookup if not built or possibly out-of-date
+            if (_cachedRulesById == null || !_cachedRulesById.TryGetValue(ruleId, out rule))
```
Comment (lgolding) [inline src/Sarif/Core/ToolComponent.cs:35]: >) [](start = 91, length = 1)

Nits:

1. you don't have to predeclare `rule`: `TryGetValue(ruleId, out ReportingDescriptor rule)`
2. you could write `if (_cachedRulesById?.TryGetValue(ruleId, out ReportingDescriptorRule) != true)`

File: src/Sarif/Readers/LineMappingStreamReader.cs [hunk 6]
```diff
@@ -11,7 +11,7 @@ namespace Microsoft.CodeAnalysis.Sarif.Readers
     ///  in the part of the file currently read. This allows it to map a line and char back to a byte offset
     ///  to allow consumers to seek back to the position later.
     /// </summary>
-    internal class LineMappingStreamReader : StreamReader
```
Comment (lgolding) [inline src/Sarif/Readers/LineMappingStreamReader.cs:14]: >internal [](start = 4, length = 8)

Why this change?

File: src/Sarif/Readers/LineMappingStreamReader.cs [hunk 7]
```diff
@@ -41,15 +41,55 @@ internal class LineMappingStreamReader : StreamReader
...
+            _bytesReadPreviously = bomWidth;
+
             // (1, 1) is the 0th byte, so the line number before the read is 1 and there is one character (the newline before the first line) before the first read.
             _firstLineNumber = 1;
             _firstLineCharsBeforeBuffer = 1;
 
             _overflowCorrector = new OverflowCorrector();
         }
 
+        private static Stream FindBomWidth(Stream stream, out int bomWidth)
```
Comment (lgolding) [inline src/Sarif/Readers/LineMappingStreamReader.cs:56]: >private [](start = 8, length = 7)

Should have unit tests (using our `testCase` pattern).

File: src/Sarif/Core/Result.cs (hunk 1 above)
Comment (ScottLouvau) [reply src/Sarif/Core/Result.cs:76]: Fixed. I'm going to leave it in. Feel free to remove GUID based matching based on how you and Mike resolve the conversation.

File: src/Sarif/Core/Result.cs (hunk 2 above)
Comment (ScottLouvau) [reply src/Sarif/Core/Result.cs:81]: Fixed

File: src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs (hunk 4 above)
Comment (ScottLouvau) [reply src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs:286]: Changed.

File: src/Sarif/Core/ToolComponent.cs (hunk 5 above)
Comment (ScottLouvau) [reply src/Sarif/Core/ToolComponent.cs:35]: I agree with #2 but I don't like inline declaration when the variable will be used in a broader context than the statement. Since this one isn't just for the if true case, I think it's weird to declare it inline, technically in an expression that won't execute (if _cachedRulesById was null) and then use it outside the whole block. I know the compiler handles it fine, but it looks weird to me.

Comment (ScottLouvau) [reply src/Sarif/Core/ToolComponent.cs:35]: Though actually I don't love #2 either. You have to think carefully about how the null boolean evaluates to the condition, whereas a X != null && .. condition is very trivial to understand.

File: src/Sarif/Readers/LineMappingStreamReader.cs (hunk 6 above)
Comment (ScottLouvau) [reply src/Sarif/Readers/LineMappingStreamReader.cs:14]: Left from running testing in another project because I couldn't get the SDK unit tests to execute. Undone.

File: src/Sarif/Readers/LineMappingStreamReader.cs (hunk 7 above)
Comment (ScottLouvau) [reply src/Sarif/Readers/LineMappingStreamReader.cs:56]: Added. I wasn't able to get tests to run, so I wasn't excited about adding them at that point.

File: src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs (hunk 3 above)
Comment (ScottLouvau) [reply src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs:280]: Yep - good catch, and added a test to cover Rule lookup by index and id and setting index and id on the result logs.
